### PR TITLE
Add tests for custom array functions in BokehJS

### LINKED ISSUE #12790: Complete unit tests for BokehJS mathematical array functions
We have a couple of BokehJS modules that contain mathematical array functions similar to equivalent-named NumPy functions:

- `bokehjs/src/lib/core/util/array.ts`
- `bokehjs/src/lib/core/util/arrayable.ts`

I recently found and fixed a couple of bugs relating to edge-case use of some of these functions in PRs #12785 and #12789. It would be good to add unit tests for all of these functions.

### PR DESCRIPTION
**Pull request by DerTimonius**
Hi everyone!

This PR will add tests for custom array functions in BokehJS that are similar to NumPy methods. The added tests will check the functions in the `bokehjs/src/lib/core/util/array.ts` file.
I also found an unnecessary `for` loop in the `repeat` function which I changed to the built-in `fill()` method.

*This PR is still work in progress, some tests are missing!*

- [x] issues: addresses #12790 
- [x] tests added / passed
- [ ] release document entry (if new feature or API change)

### CODE REVIEW

**Comment by bryevdv** [comment]
Thanks for the PR @DerTimonius ! BTW I was meaning to ask, would you like to be added to @bokeh/dev ? Mostly the group is used to solicit wider opinions on issues or PRs, also PRs can be opened against the main repo directly.

**Comment by DerTimonius** [comment]
@bryevdv Thank you for the offer, I would like that very much 😊

**Comment by DerTimonius** [comment]
Also, I'm trying to figure out why the tests aren't failing locally. I guess they are not running correctly, will figure it out next week. Will add new tests either way!

File: bokehjs/test/unit/core/util/array.ts [hunk 1]
```diff
@@ -109,4 +109,42 @@ describe("core/util/array module", () => {
...
+  })
+
+  it("remove_at should remove an item from an array at the specified index", () => {
+    expect(array.remove_at([1, 2, 3, 4], 1)).to.be.equal([1, 3, 4])
+    expect(array.remove_at([1, 2, 3, 4], 5)).to.be.equal([1, 2, 3, 4])
+    expect(array.remove_at([NaN], NaN)).to.be.equal([])
+  })
+
+  it("clear should return an empty array", () => {
+    expect(array.clear([1, 2, 3, 4])).to.be.equal([])
```
Comment (mattpap) [inline bokehjs/test/unit/core/util/array.ts]: It should not. The return type of `clear()` function is `void`.

File: bokehjs/test/unit/core/util/array.ts [hunk 2]
```diff
@@ -109,4 +109,42 @@ describe("core/util/array module", () => {
     expect(array.split([1, null, 2], null)).to.be.equal([[1], [2]])
     expect(array.split([0, 1, null, 2, null, 3, 4, 5], null)).to.be.equal([[0, 1], [2], [3, 4, 5]])
   })
+
+  it("linspace should return an array of a given length between two given numbers", () => {
+    expect(array.linspace(2, 3, 5)).to.be.equal([ 2, 2.25, 2.5, 2.75, 3 ])
+    expect(array.linspace(2, 3).length).to.be.equal(100)
+    expect(array.linspace(2, 3, 0)).to.be.equal([])
+    expect(array.linspace(2, 3, NaN)).to.be.instanceof(RangeError)
```
Comment (mattpap) [inline bokehjs/test/unit/core/util/array.ts]: This would work if `linspace()` returned `RangeError`, but it throws it, so one should use this:
```ts
expect(() => array.linspace(2, 3, NaN)).to.throw(RangeError)
```

File: bokehjs/test/unit/core/util/array.ts [hunk 3]
```diff
@@ -109,4 +109,42 @@ describe("core/util/array module", () => {
     expect(array.split([1, null, 2], null)).to.be.equal([[1], [2]])
     expect(array.split([0, 1, null, 2, null, 3, 4, 5], null)).to.be.equal([[0, 1], [2], [3, 4, 5]])
   })
+
+  it("linspace should return an array of a given length between two given numbers", () => {
+    expect(array.linspace(2, 3, 5)).to.be.equal([ 2, 2.25, 2.5, 2.75, 3 ])
```
Comment (mattpap) [inline bokehjs/test/unit/core/util/array.ts]:
```suggestion
    expect(array.linspace(2, 3, 5)).to.be.equal([2, 2.25, 2.5, 2.75, 3])
```

File: bokehjs/test/unit/core/util/array.ts [hunk 4]
```diff
@@ -109,4 +109,42 @@ describe("core/util/array module", () => {
...
+
+  it("reversed should return the given array in reversed order", () => {
+    expect(array.reversed([1, 2, 3, 4])).to.be.equal([4, 3, 2, 1])
+    expect(array.reversed([null, NaN, undefined, 0, 1])).to.be.equal([1, 0, undefined, NaN, null])
+  })
+
+  it("repeat should create an array of a given length with repeated values", () => {
+    expect(array.repeat(4, 3)).to.be.equal([4, 4, 4])
+    expect(array.repeat([1, 2], 3)).to.be.equal([[1, 2], [1, 2], [1, 2]])
+    expect(array.repeat(NaN, undefined)).to.be.equal([NaN])
```
Comment (mattpap) [inline bokehjs/test/unit/core/util/array.ts]: Repeat does not accept `undefined` for `n` argument. There is no point in testing configurations that don't type-check.

File: bokehjs/test/unit/core/util/array.ts [hunk 5]
```diff
@@ -109,4 +109,42 @@ describe("core/util/array module", () => {
...
+    expect(array.transpose([[1, 2], [3, 4]])).to.be.equal([[1, 3], [2, 4]])
+    expect(array.transpose([[1, 2, 3, 4], [5, 6, 7, 8]])).to.be.equal([[1, 5], [2, 6], [3, 7], [4, 8]])
+    expect(array.transpose([[1, 2, 3, 4], [5, 6, 7]])).to.be.equal([[1, 5], [2, 6], [3, 7], [4, undefined]])
+    expect(array.transpose([[1, null, 3, undefined], [NaN]])).to.be.equal([[1, NaN], [null, undefined], [3, undefined], [undefined, undefined]])
+  })
+
+  it("remove_at should remove an item from an array at the specified index", () => {
+    expect(array.remove_at([1, 2, 3, 4], 1)).to.be.equal([1, 3, 4])
+    expect(array.remove_at([1, 2, 3, 4], 5)).to.be.equal([1, 2, 3, 4])
+    expect(array.remove_at([NaN], NaN)).to.be.equal([])
```
Comment (mattpap) [inline bokehjs/test/unit/core/util/array.ts]: The second argument is supposed to be an integer index. The type system doesn't distinguish integers from floats, so technically this is possible, but it's hardly practical. If you want to be strict about such cases, then I would add an assertion (`assert(isInteger(i) && i >= 0)`) to `remove_at()`, and change this test to catch `AssertionError`.

File: bokehjs/test/unit/core/util/array.ts (hunk 2 above)
Comment (DerTimonius) [reply bokehjs/test/unit/core/util/array.ts]: Thank you for your input (here and on the other chunks of code)! I was looking for something like `to.throw` but could not find it..I guess my brian is a bit too wired using Jest and Playwright 😄

File: bokehjs/test/unit/core/util/array.ts [hunk 6]
```diff
@@ -111,29 +112,27 @@ describe("core/util/array module", () => {
...
     expect(array.remove_at([1, 2, 3, 4], 1)).to.be.equal([1, 3, 4])
     expect(array.remove_at([1, 2, 3, 4], 5)).to.be.equal([1, 2, 3, 4])
-    expect(array.remove_at([NaN], NaN)).to.be.equal([])
+    expect(() => array.remove_at([NaN], NaN)).to.throw(AssertionError)
   })
 
-  it("clear should return an empty array", () => {
-    expect(array.clear([1, 2, 3, 4])).to.be.equal([])
+  it("clear should return nothing", () => {
+    expect(array.clear([1, 2, 3, 4])).to.be.equal(undefined)
```
Comment (mattpap) [inline bokehjs/test/unit/core/util/array.ts]: Actually, this is not what I meant with me previous comment. The test should verify that this function actually does anything useful, something like:
```ts
const arr = [1, 2, 3, 4]
array.clear(arr)
expect(arr).to.be.equal([]) // or
expect(arr.length).to.be.equal(0)
```
Given that we heavily rely to TypeScript, then testing whether void function returns `undefined` is unnecessary.

**Comment by DerTimonius** [comment]
I was wondering if other functions are in need of specific tests @mattpap?

**Comment by mattpap** [comment]
> I was wondering if other functions are in need of specific tests 

Every function should have at least some unit tests. If you want to contribute more, that's great. Such contributions can be submitted in batches (separate PRs), so let us know how you want to proceed. We can retarget this PR for 3.1 if you consider it ready (just needs a rebase).

**Comment by DerTimonius** [comment]
I think I would proceed with finishing this one first (resolving the merge conflict) later today and add more tests in separate PRs 👍

**Review by mattpap — APPROVED** [review]
